zvariant/src/file_path.rs

```diff
@@ -0,0 +1,348 @@
+///    let p7 = FilePath::from("/hello/world");
+///
+///    assert_eq!(p1, p2);
+///    assert_eq!(p2, p3);
+///    assert_eq!(p3, p4);
+///    assert_eq!(p4, p5);
+///    assert_eq!(p5, p6);
+///    assert_eq!(p5, p7);
+/// ```
+/// Also you can (de)serialize the [FilePath] as an array of bytes, consider this for example
+///
+/// ```
+///    use zvariant::FilePath;
+///    use serde_json::Value::Array;
+///    use serde::Deserialize;
+///    use std::borrow::Cow;
+///    use std::ffi::CStr;
+///    use std::path::PathBuf;
+///
+///    let file_path = FilePath::from("/hello/world");
+///    let path_ser = serde_json::json!(file_path);
+///     // value = /hell/world
+///    let path_arr = Array(vec![
+///        serde_json::json!(47),
+///        serde_json::json!(104),
+///        serde_json::json!(101),
+///        serde_json::json!(108),
+///        serde_json::json!(108),
+///        serde_json::json!(111),
+///        serde_json::json!(47),
+///        serde_json::json!(119),
+///        serde_json::json!(111),
+///        serde_json::json!(114),
+///        serde_json::json!(108),
+///        serde_json::json!(100),
+///    ]);
+///    assert_eq!(path_ser, path_arr);
+///
+///    let path_deser: FilePath<'static> = FilePath::deserialize(path_ser).unwrap();
+///    assert_eq!(file_path, path_deser);
+///
+///    let path_buf: PathBuf = file_path.into();
+///    let string_lossy = Cow::Borrowed("/hello/world");
+///    let path_lossy = path_buf.to_string_lossy();
+///    assert_eq!(string_lossy, path_lossy);
+///
+///     // /hello/world/\\xC9blahblah
+///    let x = [
+///        47, 104, 101, 108, 108, 111, 47, 119, 111, 114, 108, 100, 47,
+///        92, 120, 67, 57, 98, 108, 97, 104, 98, 108, 97, 104, 0x0,
+///    ];
+///    let file_path = FilePath::new(Cow::Borrowed(CStr::from_bytes_with_nul(&x).unwrap()));
+///    let p2 = FilePath::from(PathBuf::from("/hello/world/\\xC9blahblah"));
+///    assert_eq!(p2, file_path);
+/// ```
+#[derive(Type, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+#[zvariant(signature = "ay")]
+pub struct FilePath<'f>(Cow<'f, CStr>);
+
+impl<'f> FilePath<'f> {
+    pub fn new(cow: Cow<'f, CStr>) -> Self {
+        Self(cow)
+    }
+}
+
+/// Chops a vector of bytes until a nul character,
+/// and if it (the nul) does not exist, it appends it at
+/// the end.
+///
+/// # Return:
+/// - _*Guaranteed*_ nul terminated vector of bytes
+fn chop_vec_with_nul(vec: &Vec<u8>) -> Vec<u8> {
+    let mut vec_with_nul: Vec<u8> = vec![];
+    for v in vec {
+        if v == &0x0 {
+            vec_with_nul.push(v.to_owned());
+            return vec_with_nul;
+        }
+        vec_with_nul.push(v.to_owned());
+    }
+    vec_with_nul.push(0x0);
+    return vec_with_nul;
+}
+
+/// Chops a slice of bytes until a nul character,
+/// and if it (the nul) does not exist, it appends it at
+/// the end.
+/// This method won't allocate/copy memory unless the nul
+/// byte does not exist in the `bytes` parameter
```

Comment:
I don't think we need to do all this for the user. We should just simply check for the null byte and error out if it's not there.